### PR TITLE
Modify PhasedActions to run tasks only when explicitly requested by the client

### PR DESCRIPTION
**Pull request by lsmaira**
Modify Phased Actions to run tasks only when defined by the client and, otherwise
just configure the build.

projectsEvaluated hook is removed and projectsLoaded is executed (for now) once projects
are configured.

Signed-off-by: Lucas Smaira <lsmaira@google.com>

### Context
This is a follow up for PR https://github.com/gradle/gradle/pull/4533

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation
- [x] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/initialization/DefaultGradleLauncherFactory.java [hunk 1]
```diff
@@ -283,6 +283,11 @@ public GradleInternal configure() {
                 throw new UnsupportedOperationException();
             }
 
+            @Override
+            public GradleInternal build(Stage upTo) {
+                throw new UnsupportedOperationException();
```
Comment (lsmaira) [inline subprojects/core/src/main/java/org/gradle/initialization/DefaultGradleLauncherFactory.java]: Is this correct?

**Comment by lsmaira** [comment]
Thank you very much for reviewing the other big PR, @oehme ! I hope this one will be shorter :)

I think there are two possible approaches concerning the new API in BuildController:
- Either we make "Projects Loaded" as a separate BuildOperation from "ProjectConfiguration" and then we are able to stop after that
- Or we make "Projects Loaded" as part of "Load Build" BuildOperation.

Is that correct?

**Comment by oehme** [comment]
I thought about this some more and I think we don't actually need the `upToPhase` API, but can instead make the following changes to the existing feature:

- if the user specified any tasks (using `withTasks`), run tasks, otherwise configure the build, just like the already existing action handlers do
- if you want to determine tasks yourself in a model builder, you need to add `withTasks(emptyList)` to tell Gradle that you intend to run tasks
- remove `projectsEvaluated` and only keep `projectLoaded` and `buildFinished`. Change the documentation of `projectsLoaded` to say that this is called when the projects are loaded and that they will be configured as necessary for the models requested. In reality we will still configure everything for now, but it leaves a door open for us to configure less in the future
- rename `phasedAction()` to just `action()`

I think this solves your use case without introducing additional complexity and it exposes less of the "phase" concepts which we are trying to steer away from. For instance, with configure-on-demand, projects can be configured very late, so a `projectsConfigured` hook could lead to confusion.

Does this sound good to you? If so, could you make those adjustments by Monday so we can get them into 4.8? Yes, it's already release time again :)

**Comment by lsmaira** [comment]
@oehme thank you for looking into that! Sure, I can do what you suggested by Monday (I'll try to send it to you later today).

This will solve our use case, however, let me know in the future if you have a hint of how to avoid unnecessary configuration and I can implement it later.

**Comment by oehme** [comment]
Regarding avoiding configuration: The idea would be that we only configure the projects for which models are requested and that a model builder can say "to use me, you also need to configure subprojects", so that things like the hierarchical `EclipseProject` would keep working. Of course this is pretty handwavy and we'll have some subtleties to work out :)

**Comment by lsmaira** [comment]
@oehme I made the changes you suggested and run the tests locally. Please, let me know if docs are good enough and if there's anything missing.
Also, I didn't get if I should rename only the method phasedAction() to action() or all the classes, etc. Please, let me know

**Comment by oehme** [comment]
Just the method, the class names can stay the same.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/PhasedBuildActionExecuter.java [hunk 2]
```diff
@@ -86,18 +73,20 @@
     }
 
     /**
-     * Specifies the tasks to execute before executing the BuildFinishedAction and after the ProjectsEvaluatedAction.
+     * Specifies the tasks to execute before executing the BuildFinishedAction and after the ProjectsLoadedAction.
      *
      * @param tasks The paths of the tasks to be executed. Relative paths are evaluated relative to the project for which this launcher was created.
+     * It can be an empty collection to indicate that tasks should be run.
```
Comment (oehme) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/PhasedBuildActionExecuter.java]: Passing an empty collection will run the default tasks

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/PhasedBuildActionExecuter.java [hunk 3]
```diff
@@ -86,18 +73,20 @@
...
      */
     @Incubating
     PhasedBuildActionExecuter forTasks(String... tasks);
 
     /**
-     * Specifies the tasks to execute before executing the BuildFinishedAction and after the ProjectsEvaluatedAction.
+     * Specifies the tasks to execute before executing the BuildFinishedAction and after the ProjectsLoadedAction.
      *
      * @param tasks The paths of the tasks to be executed. Relative paths are evaluated relative to the project for which this launcher was created.
+     * It can be an empty collection to indicate that tasks should be run.
```
Comment (oehme) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/PhasedBuildActionExecuter.java]: Same as above

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/DefaultPhasedBuildActionExecuter.java [hunk 4]
```diff
@@ -47,13 +49,13 @@ protected DefaultPhasedBuildActionExecuter getThis() {
 
     @Override
     public PhasedBuildActionExecuter forTasks(String... tasks) {
-        operationParamsBuilder.setTasks(rationalizeInput(tasks));
+        operationParamsBuilder.setTasks(tasks != null ? Arrays.asList(tasks) : null);
```
Comment (oehme) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/DefaultPhasedBuildActionExecuter.java]: Why this change? The method already does what you want.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/DefaultPhasedBuildActionExecuter.java [hunk 5]
```diff
@@ -47,13 +49,13 @@ protected DefaultPhasedBuildActionExecuter getThis() {
...
         return getThis();
     }
 
     @Override
     public PhasedBuildActionExecuter forTasks(Iterable<String> tasks) {
-        operationParamsBuilder.setTasks(rationalizeInput(tasks));
+        operationParamsBuilder.setTasks(tasks != null ? CollectionUtils.toList(tasks) : null);
```
Comment (oehme) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/DefaultPhasedBuildActionExecuter.java]: Same as above

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/DefaultPhasedBuildActionExecuter.java (hunk 4 above)
Comment (oehme) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/DefaultPhasedBuildActionExecuter.java]: I missed that, sorry.

**Comment by lsmaira** [comment]
Not sure this build failure was related to the code, I tried rebasing...

**Comment by oehme** [comment]
Thanks again @lsmaira! This will be released with 4.8
